geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/SSLDUnitTest.java

```diff
@@ -0,0 +1,298 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.io.FileInputStream;
+import java.security.KeyStore;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Properties;
+
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLHandshakeException;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.TrustManagerFactory;
+
+import io.netty.handler.ssl.NotSslRecordException;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.exceptions.JedisConnectionException;
+
+import org.apache.geode.cache.ssl.CertStores;
+import org.apache.geode.cache.ssl.CertificateBuilder;
+import org.apache.geode.cache.ssl.CertificateMaterial;
+import org.apache.geode.internal.net.filewatch.PollingFileWatcher;
+import org.apache.geode.test.awaitility.GeodeAwaitility;
+import org.apache.geode.test.dunit.IgnoredException;
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.dunit.rules.RedisClusterStartupRule;
+
+public class SSLDUnitTest {
+
+  @Rule
+  public RedisClusterStartupRule cluster = new RedisClusterStartupRule();
+
+  private static String commonPassword = "password";
+  private static SANCapturingHostnameVerifier hostnameVerifier = new SANCapturingHostnameVerifier();
+
+  private int redisPort;
+  private CertificateMaterial ca;
+  private String serverKeyStoreFilename;
+  private String serverTrustStoreFilename;
+
+  @Before
+  public void setup() throws Exception {
+    ca = new CertificateBuilder()
+        .commonName("Test CA")
+        .isCA()
+        .generate();
+
+    CertificateMaterial serverCertificate = new CertificateBuilder()
+        .commonName("server")
+        .issuedBy(ca)
+        .generate();
+
+    CertStores serverStore = CertStores.serverStore();
+    serverStore.withCertificate("server", serverCertificate);
+    serverStore.trust("ca", ca);
+    Properties serverProperties = serverStore.propertiesWith("all", true, false);
+
+    MemberVM server = cluster.startRedisVM(0, x -> x.withProperties(serverProperties));
+
+    redisPort = cluster.getRedisPort(server);
+    serverKeyStoreFilename = serverProperties.getProperty("ssl-keystore");
+    serverTrustStoreFilename = serverProperties.getProperty("ssl-truststore");
+  }
+
+  @Test
+  public void givenMutualAuthentication_clientCanConnect() throws Exception {
+    try (Jedis jedis = createClient(true, false)) {
+      assertThat(jedis.ping()).isEqualTo("PONG");
+    }
+  }
+
+  @Test
+  public void givenMutualAuthentication_clientErrorsWithoutKeystore() throws Exception {
```

Comment:
I may just be misunderstanding, but this test name implies mutual authentication, but the boolean `mutualAuthentication` argument passed to the `createClient()` method is false.

Reply:
So the `givenMutualAuthentication` is supposed to refer to how the server is configured. Passing `false` to `createClient` in order to _not_ set it up to perform mutual auth and thus error.

Reply:
Maybe the test name could be "givenMutualAuthenticationOnServer..."? Not super important though.